homeassistant/components/light/tcpbulbs.py

```diff
@@ -0,0 +1,146 @@
+    vol.Required(CONF_HOST): cv.string,
+})
+
+
+def pullXML(host, bulbxml):
+    REQUIREMENTS = ['xmltodict']
+    import xmltodict
+    resp = tcpAction(host, 0, 'xml', 0)
+    # Store bulb information as xml file
+    with open(bulbxml, 'wb') as f:
+        f.write(resp.content)
+
+    with open(bulbxml) as fd:
+        doc = xmltodict.parse(fd.read())
+        return doc
+
+def numMap(value, leftMin, leftMax, rightMin, rightMax):
+    # Figure out how 'wide' each range is
+    leftSpan = leftMax - leftMin
+    rightSpan = rightMax - rightMin
+
+    # Convert the left range into a 0-1 range (float)
+    valueScaled = float(value - leftMin) / float(leftSpan)
+
+    # Convert the 0-1 range into a value in the right range.
+    return int(rightMin + (valueScaled * rightSpan))
+
+def getBrightness(device):
+        if 'level' in device:
+            level = numMap(int(device['level']),1,100,1,255)
+            return(level)
+        else:
+            return(0)
+            
+def checkOnline(device):
+        if 'offline' in device:
+            return(False)
+        else:
+            return(True)
+
+def tcpAction(host, did, action, value):
+    import requests
+    if action == 'xml':
+        url = 'http://' + host + '/gwr/gop.php?cmd=GWRBatch&data=<gwrcmds><gwrcmd><gcmd>RoomGetCarousel</gcmd><gdata><gip><version>1</version><token>1234567890</token><fields>name,status</fields></gip></gdata></gwrcmd></gwrcmds>&fmt=xml'
+        resp = requests.get(url)
+        return resp
+    if action == 'level':
+        action = '<type>level</type>'
+    else:
+        action = ''
+    url = 'http://' + host + '/gwr/gop.php?cmd=DeviceSendCommand&data=<gip><version>1</version><token>1234567890</token><did>' + did + '</did><value>' + value + '</value>' + action + '</gip>&fmt=xml'
+    requests.get(url)       
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+
+    # Assign configuration variables. The configuration check takes care they are
+    # present. 
+    host = config.get(CONF_HOST)
+    bulbxml = hass.config.path('tcpbulbs.xml')
+    doc = pullXML(host,bulbxml)
+        
+    add_devices(TcpLights(device,host,bulbxml) for device in doc['gwrcmds']['gwrcmd']['gdata']['gip']['room']['device'])
+
+class TcpLights(Light):
+    
+    """Representation of an TCP Connected Light."""
+
+    def __init__(self, light, host, bulbxml):
+        """Initialize a TCP Connected Light."""
+        self._did = light['did']
+        self._name = light['name']
+        self._state = int(light['state'])
+        self._brightness = getBrightness(light)
+        self._host = host
+        self._bulbxml = bulbxml
+        self._online = checkOnline(light)
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return SUPPORTED_FEATURES
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self._online
+    
+    @property
+    def did(self):

```

Comment:
Why not name this also `device_id`? It would be cleaner and more understandable that way.

Reply:
Just because that's how the gateway named it. I can change that.